### PR TITLE
fixes crate animation breaking at closing

### PR DESCRIPTION
**Pull request by MNarath1**
## About The Pull Request

Reason for the small blip in the crate animation was that originally i assumed the animation parameter would be inverted mathematically indifferent but that was not the case cause of a missmatching index that happened trough this was solveable with some additional logic checks so the list size stays the same actually shrinks cause i also fixed another issue that had me add one additional index to avoid a runtime.
## Why It's Good For The Game

Fixes bumping crate animation

## Changelog
:cl:
fix: crate animation at closing
refractor: decreases crate list size by 1 deleting an unused index
tweak: changing crate animation time from 2 to 3
/:cl:

### CODE REVIEW

**Comment by MNarath1** [comment]
@park66665 this is the PR to fix the blip in crate animation i am probably going to investigate the problem further later on by exporting the whole math into matlab and looking at it step by step but for now this seems to be the only way to guarantee it will work even with a different time.

**Comment by MNarath1** [comment]
ok i am trying one last thing beforehand give me a few minutes problem is a single index is floating with one getting lost and the last one beeing replaced with another index basically

**Comment by MNarath1** [comment]
i am just about to finish the commit for it if it works should have the result in like 1-2 minutes

File: code/game/objects/structures/crates_lockers/crates.dm [hunk 1]
```diff
@@ -71,10 +71,9 @@
...
 	for(var/I in 0 to num_steps)
-		var/angle = door_anim_angle * (closing ? 1 - (I/num_steps) : (I/num_steps))
-		var/door_state = angle >= 90 ? "[icon_door_override ? icon_door : icon_state]_back" : "[icon_door || icon_state]_door"
-		var/door_layer = angle >= 90 ? FLOAT_LAYER : ABOVE_MOB_LAYER
-		var/matrix/M = get_door_transform(animation_math_list[closing ? num_steps + 1 - I : I + 1], animation_math_list[closing ? 2 * num_steps + 1 - I : num_steps + I + 1])
+		var/door_state = I == 0 || (I == num_steps && closing) ? "[icon_door || icon_state]_door" : animation_math_list[closing ? 2 * num_steps - I : num_steps + I] <= 0 ? "[icon_door_override ? icon_door : icon_state]_back" : "[icon_door || icon_state]_door"
+		var/door_layer = I == 0 || (I == num_steps && closing) ? ABOVE_MOB_LAYER : animation_math_list[closing ? 2 * num_steps - I : num_steps + I] <= 0 ? FLOAT_LAYER : ABOVE_MOB_LAYER
+		var/matrix/M = get_door_transform(I == 0 || (I == num_steps && closing) ? 0 : animation_math_list[closing ? num_steps - I : I], I == 0 || (I == num_steps && closing) ? 1 : animation_math_list[closing ?  2 * num_steps - I : num_steps + I])
```
Comment (park66665) [inline code/game/objects/structures/crates_lockers/crates.dm]:
```suggestion
		var/door_state = I == (closing ? num_steps : 0) ? "[icon_door || icon_state]_door" : animation_math_list[closing ? 2 * num_steps - I : num_steps + I] <= 0 ? "[icon_door_override ? icon_door : icon_state]_back" : "[icon_door || icon_state]_door"
		var/door_layer = I == (closing ? num_steps : 0) ? ABOVE_MOB_LAYER : animation_math_list[closing ? 2 * num_steps - I : num_steps + I] <= 0 ? FLOAT_LAYER : ABOVE_MOB_LAYER
		var/matrix/M = get_door_transform(I == (closing ? num_steps : 0) ? 0 : animation_math_list[closing ? num_steps - I : I], I == (closing ? num_steps : 0) ? 1 : animation_math_list[closing ?  2 * num_steps - I : num_steps + I])
```

Looks like you need an exception handling for the frame when the crate is completely closed. `I == (closing ? num_steps : 0)` is a correct way to check it.

File: code/game/objects/structures/crates_lockers/crates/critter.dm [hunk 2]
```diff
@@ -43,7 +43,7 @@
 	var/num_steps = door_anim_time / world.tick_lag
 	var/list/animation_math_list = animation_math["[door_anim_time]-[door_anim_angle]-[azimuth_angle_2]-[radius_2]-[door_hinge]"]
 	for(var/I in 0 to num_steps)
-		var/matrix/M = get_door_transform(animation_math_list[closing ? num_steps + 1 - I : I + 1], animation_math_list[closing ? 2 * num_steps + 1 - I : num_steps + I + 1])
+		var/matrix/M = get_door_transform(I == 0 || (I == num_steps && closing) ? 1 : animation_math_list[closing ? num_steps - I : I], I == 0 || (I == num_steps && closing) ? 0 : animation_math_list[closing ? 2 * num_steps - I : num_steps + I])
```
Comment (park66665) [inline code/game/objects/structures/crates_lockers/crates/critter.dm]:
```suggestion
		var/matrix/M = get_door_transform(I == (closing ? num_steps : 0) ? 1 : animation_math_list[closing ? num_steps - I : I], I == (closing ? num_steps : 0) ? 0 : animation_math_list[closing ? 2 * num_steps - I : num_steps + I])
```

ditto

File: code/game/objects/structures/crates_lockers/crates/critter.dm [hunk 3]
```diff
@@ -81,8 +81,8 @@
 /obj/structure/closet/crate/critter/animation_list()
 	var/num_steps_1 = door_anim_time / world.tick_lag
 	var/list/new_animation_math_sublist[num_steps_1 * 2 + 1]
```
Comment (park66665) [inline code/game/objects/structures/crates_lockers/crates/critter.dm]:
```suggestion
	var/list/new_animation_math_sublist[num_steps_1 * 2]
```

**Review by park66665 — COMMENTED** [review]
There also is a small visual bug (some crates' opening animation gets a bit weird at first few frames), but I doubt that can be fixed: if that bug really can be fixed, I bet it'll require a complete rewrite of the crate animation code.

But I have a proposal. Increase crates' `door_anim_time` to 3. That'll mitigate the visual bug and make crate animation better imho.

**Review by park66665 — APPROVED** [review]
Animation works correctly now. Let's wait to see if anyone opens a counter-PR. YOLO

**Pull request merged**